### ISSUE TITLE
Energy Secondary Data Holder & Application Specific Errors

### ISSUE THREAD

**Issue by OpalRussAEMO** [ConsumerDataStandardsAustralia/standards-maintenance#478]
# Description
AEMO, as secondary data holder, has identified several scenarios under which data cannot be returned for shared data responsibility energy APIs.  A solution on how these scenarios should be communicated needs to be determined.

To preface this, AEMO expects energy retailers (as primary data holders), will in most cases prevent requests from coming to AEMO for these scenarios.  In the event that there may be a data discrepancy between a retailer's system and AEMO's, AEMO will still need to validate for these conditions.

Scenarios include:

1. getServicePointDetail - Invalid NMI Status

- Extinct : service point has been abolished and data will no longer be provided for the service point. 
- Off Market : an embedded network child service point previously on market, has switched to off market, AEMO are no longer the source of information for the service point.  If data is required, it will need to be requested via the embedded network manager.
- Greenfield: an identifier (NMI) for the service point has been established however the connection and / or metering is not yet complete.  Data may be available at a later date once the connection and metering is complete.

2. All SDH APIs - Requesting participant is not the registered FRMP

- Service Point & DER APIs: participant ID provided does not match the current registered FRMP Participant ID.
- Energy Usage: participant ID provided is not registered as FRMP for any period of the requested date range.  

3. getServicePointDetail - No active meter or register records

- GetServicePointDetail requires at least one meter and one register record.  If no meters or registers are active then data cannot be returned.

# Area Affected
The high level standards / Error Codes section.
https://consumerdatastandardsaustralia.github.io/standards/#error-codes

# Change Proposed
There are multiple options.  The solution may be one or a combination depending on how the above scenarios are to be handled.

**Option 1:** Use existing 422 errors: either InvalidNMI or UnavailableNMI
**Option 2:** Introduce new error codes in the standards for the listed scenarios
**Option 3:** Do not return any error nor any data. 
**Option 4:** Application specific error codes to be returned by secondary data holder and pass through supported by primary data holder
**Option 5:** To cater for scenario groups 1 & 3: update meter and register details to be optional on getServicePointDetails.   ADR will receive sufficient information to be able to identify these scenarios themselves.

**Comment by OpalRussAEMO** [comment]
AEMO's preference is:

- **Scenario groups 1 & 3** : implement option 5, update meter and register details to optional and only provided where available and current.  This will allow data recipients to receive the status of the service point which is sufficient information to identify the scenario.  AEMO will provide meter and register details in all cases where available and current.

- **Scenario group 2**: option 2, introduce a new error code where participant is not entitled to data

**Comment by CDR-API-Stream** [comment]
The DSB and AEMO have had further discussions on this and recommend the following approach:

- **Scenario groups 1 & 3** : The DSB agrees with the recommendation made by AEMO to consider option 5 and update “meters” and “registers” objects within [EnergyServicePointDetail](https://consumerdatastandardsaustralia.github.io/standards/#tocSenergyservicepointdetail) schema to optional so they can be provided where available. 

- **Scenario group 2**: No change to the standards is required to cater for this as per below explanation:

  -  **Service Point & DER APIs: participant ID provided does not match the current registered FRMP Participant ID**. : No change required as this is a standard scenario.  Appropriate error code (e.g. 404 or 422) would be used based on the scenario. 
  - **Energy Usage: participant ID provided is not registered as FRMP for any period of the requested date range.**: This would result in a valid response from AEMO with usage data covering the period for which the retailer for the FRMP would be returned.  In an event where retailer is the current FRMP of the service point but was not the FRMP for the entire requested period, an empty array would be returned.  Consequently, no change to the standards are required.

**Comment by perlboy** [comment]
Firstly, it's getting very difficult to provide comment on Secondary Data Holder APIs if the proposed API isn't _actually_ published. Reading between the lines of how this will work is not conducive to effective review.

Nonetheless, if a NMI is Extinct does this mean that usage data won't be retrievable either or _just_ the detailed information regarding the NMI?

**Comment by perlboy** [comment]
Further, since there is discussion around application specific errors. AEMO currently has a series of error codes and I believe is proposing the use of these as application specific codes (`InvalidSP-xxx`) while decreasing the accuracy of the CDR Standards error code (ie. a single CDR error code with many application specific codes)? If so, this is problematic because:
1) It requires an ADR to process Application Specific Code's to ascertain the reason for a failure
2) It precludes a primary Data Holder from using Application Specific Code's themselves

The Standards currently state:
> Where an application supports error codes specific to that implementation, it is intended that implementations can **extend the standard CDR error codes** with application-specific error responses whilst **maintaining interoperability** for clients.

Baking in Application Error Codes for an ecosystem wide Secondary Data Holder (ie. AEMO) is, in essence, forcing those to become the standardised behaviour for Energy participants and go far beyond the concept of "extend". This would effectively diminish the authority that the Standards themselves have in this regard.

Suggest that, at a minimum, AEMO have a 1:1 set of error codes between Standards and Application specific codes and, since AEMO is a government body itself presumably intending to embrace the CDR process, preferably abandon application specific codes for CDR operations entirely and explicitly align with the CDR Standards by adopting Standards Error Codes with sufficient definition.

**Comment by OpalRussAEMO** [comment]
> Firstly, it's getting very difficult to provide comment on Secondary Data Holder APIs if the proposed API isn't _actually_ published. Reading between the lines of how this will work is not conducive to effective review.
> 
> Nonetheless, if a NMI is Extinct does this mean that usage data won't be retrievable either or _just_ the detailed information regarding the NMI?

Issue with extinct NMIs is that they don't have any meters or registers (as supply point is abolished), changing meter/register objects to optional means we can provide the service point details for Extinct NMIs.

AEMO will provide any meter data that is available for a period where the requesting party is responsible regardless of the NMI status.  

With the exception of the header, the API request / response structure is as published by the DSB.

**Comment by perlboy** [comment]
Thanks for the quick response @OpalRussAEMO 

> Issue with extinct NMIs is that they don't have any meters or registers (as supply point is abolished), changing meter/register objects to optional means we can provide the service point details for Extinct NMIs.

Totally understand and makes sense, my concern is around consumers requesting historical data but it not being available if the meter was extinct (ie. house redevelopment). In the extinct scenario for Get Service Points who is the FRMP because at this point AEMO is restricting such calls to the FRMP of the service point?

> AEMO will provide any meter data that is available for a period where the requesting party is responsible regardless of the NMI status.

This is in conflict with the unpublished business rules.

> With the exception of the header, the API request / response structure is as published by the DSB.

This is not true. 

There exists an AEMO document which outlines a series of business rules, error behaviours and attribute overrides (`servicePointId`) which conflict with the guidance above. This document is covered by "the copyright permissions on AEMO's website" which only state that distribution is permitted when _["made publicly available by AEMO"](https://aemo.com.au/en/privacy-and-legal-notices/copyright-permissions)_.

Please make this document publicly available.

**Comment by OpalRussAEMO** [comment]
> Further, since there is discussion around application specific errors. AEMO currently has a series of error codes and I believe is proposing the use of these as application specific codes (`InvalidSP-xxx`) while decreasing the accuracy of the CDR Standards error code (ie. a single CDR error code with many application specific codes)? If so, this is problematic because:
> 
> 1. It requires an ADR to process Application Specific Code's to ascertain the reason for a failure
> 2. It precludes a primary Data Holder from using Application Specific Code's themselves
> 
> The Standards currently state:
> 
> > Where an application supports error codes specific to that implementation, it is intended that implementations can **extend the standard CDR error codes** with application-specific error responses whilst **maintaining interoperability** for clients.
> 
> Baking in Application Error Codes for an ecosystem wide Secondary Data Holder (ie. AEMO) is, in essence, forcing those to become the standardised behaviour for Energy participants and go far beyond the concept of "extend". This would effectively diminish the authority that the Standards themselves have in this regard.
> 
> Suggest that, at a minimum, AEMO have a 1:1 set of error codes between Standards and Application specific codes and, since AEMO is a government body itself presumably intending to embrace the CDR process, preferably abandon application specific codes for CDR operations entirely and explicitly align with the CDR Standards by adopting Standards Error Codes with sufficient definition.

We've agreed we won't use application specific errors.  This issue covers all data issue scenarios that we have identified.  We will be looking to implement as per the DSB recommendation above.

We did have application specific errors in the guide for what AEMO is implementing.  That was just a draft document for discussion and not finalised.  We'll be removing them.

**Comment by perlboy** [comment]
> We did have application specific errors in the guide for what AEMO is implementing. That was just a draft document for discussion and not finalised. We'll be removing them.

Discussion behind closed doors to selected parties is not an open consultation. Publishing post "finalisation" is neither in the spirit of CDR nor advisable as it will, in all likelihood, be done so without the benefit of input from parties involved in the CDR for years.

**Comment by OpalRussAEMO** [comment]
> Thanks for the quick response @OpalRussAEMO
> 
> > Issue with extinct NMIs is that they don't have any meters or registers (as supply point is abolished), changing meter/register objects to optional means we can provide the service point details for Extinct NMIs.
> 
> Totally understand and makes sense, my concern is around consumers requesting historical data but it not being available if the meter was extinct (ie. house redevelopment). In the extinct scenario for Get Service Points who is the FRMP because at this point AEMO is restricting such calls to the FRMP of the service point?
> 
> > AEMO will provide any meter data that is available for a period where the requesting party is responsible regardless of the NMI status.
> 
> This is in conflict with the unpublished business rules.
> 
> > With the exception of the header, the API request / response structure is as published by the DSB.
> 
> This is not true.
> 
> There exists an AEMO document which outlines a series of business rules, error behaviours and attribute overrides (`servicePointId`) which conflict with the guidance above. This document is covered by "the copyright permissions on AEMO's website" which only state that distribution is permitted when _["made publicly available by AEMO"](https://aemo.com.au/en/privacy-and-legal-notices/copyright-permissions)_.
> 
> Please make this document publicly available.

We've combined the API specifications with the variances that are specified under the Shared Responsibility section on the DSB.    servicePointId needing to be populated with NationalMeteringId is covered under that section.  Aside from working through data issue scenarios and the additional headers I don't believe there is anything else that is conflicting.

**Comment by perlboy** [comment]
> We've combined the API specifications with the variances that are specified under the Shared Responsibility section on the DSB. servicePointId needing to be populated with NationalMeteringId is covered under that section. Aside from working through data issue scenarios and the additional headers I don't believe there is anything else that is conflicting.

Then you are wrong but I'm unable to explicitly spell it out because the material isn't publicly available and therefore the legality of discussing further is unclear.

**Comment by johnAEMO** [comment]
I think that this example of public consultation is inappropriate and no longer constructive - can the moderator intervene please?

**Comment by CDR-API-Stream** [comment]
We have locked this thread temporarily while we review

**Comment by CDR-API-Stream** [comment]
All feedback provided in response to consultations on the CDR data standards are expected to comply with the [GitHub Community Forum Code of Conduct](https://docs.github.com/en/github/site-policy/github-community-forum-code-of-conduct).  This code is based on the best practices of 'Be respectful and considerate' and 'Contribute in a positive and constructive way'.

After review and internal discussions it would appear that the feedback provided is not in keeping with the spirit of this best practice guidance and we request that community participants be mindful of this.  Participants are also reminded that they can contact the DSB directly (at: [contact@consumerdatastandards.gov.au](mailto:contact@consumerdatastandards.gov.au)) if they have concerns about how consultations are progressing.

We will be keeping this thread locked until we are able to discuss it at the next maintenance iteration meeting.  It will be unlocked depending on the outcome of that discussion.

**Comment by CDR-API-Stream** [comment]
Based on the conversation in the MI call this thread will now be unlocked and participants are requested keep the code of conduct in mind.

In addition, it was discussed that a OAS version of the shared responsibility endpoint, reflecting the current state of the standards will be added to v1.16.1 of the standards with any additional changes arising from this maintenance iteration being included in v1.17.0 at the end of the iteration.

**Comment by CDR-API-Stream** [comment]
Based on discussions to date it appears that the current recommendation for this issue would be:

1. Make meters optional to accommodate situations where they don't arise
2. Reuse existing errors wherever possible
3. Add standardised errors for any identified new scenarios not already covered

**Comment by CDR-API-Stream** [comment]
As per the action captured during [MI 10 meeting on 30th March](https://github.com/ConsumerDataStandardsAustralia/standards/wiki/DSB-Maintenance-Iteration-10:-Agenda-&-Meeting-Notes-(30-March-2022)), below table provides the recommendations for scenarios raised and discussed as part of this CR:

| Scenario | Impacted APIs | Treatment | Recommendation |
|:-------|:------------|:---------|:------------------|
| NMI status is either EXTINCT, GREENFIELD or OFF_MARKET | getServicePointDetail | AEMO will provide standing data where available for current or disconnected meters | Make “meters” and “register” objects in [EnergyServicePointDetail](https://consumerdatastandardsaustralia.github.io/standards/?examples#tocSenergyservicepointdetail) optional to cater for scenarios where meter or register information for a NMI is not available |
| Retailer is not the registered FRMP | getServicePoints, getServicePointDetails, getDERForServicePoint, getDERForSpecificServicePoints | This is a standard scenario. Appropriate existing error codes (e.g. 404 or 422) would be used as required | No change required |
| Retailer is the FRMP for part of the requested date range for usage data | getUsageForServicePoint | Data will be returned for periods within the requested date range where the requesting participant is the registered FRMP.  <br /><br />Where the requesting party is not the registered Financially Responsible Market Participant (FRMP) for any part of the request date range, an empty array will be returned.  No error will be included. | No change required |

**Comment by CDR-API-Stream** [comment]
This change was incorporated into [release v1.17.0](https://consumerdatastandardsaustralia.github.io/standards/includes/releasenotes/releasenotes.1.17.0.html#v1-17-0-release-notes). Refer to [Decision 237](https://github.com/ConsumerDataStandardsAustralia/standards/issues/237)  for further details.
